src/mlpack/methods/decision_tree/decision_tree_impl.hpp

```diff
@@ -656,9 +656,24 @@ void DecisionTree<FitnessFunction,
       }
       else
       {
-        child->Train<UseWeights>(data, currentChildBegin,
+        double prevChildGain = FitnessFunction::template Evaluate<UseWeights>(
+            labels.subvec(currentChildBegin, currentCol - 1),
+            numClasses,
+            UseWeights ? weights.subvec(currentChildBegin, currentCol - 1) :
+                weights);
+        // During recursion entropy of child node may change
+        double newChildGain = child->Train<UseWeights>(data, currentChildBegin,
             currentCol - currentChildBegin, datasetInfo, labels, numClasses,
             weights, minimumLeafSize, minimumGainSplit);
+        newChildGain = -newChildGain;
+        if(newChildGain>prevChildGain)
+        {
+          double prevContribute = prevChildGain * childCounts[i] / count ;
+          double newBestGain = bestGain - prevContribute;
+          double newContribute = newChildGain * childCounts[i] / count;
+          newBestGain = newBestGain + newContribute;
+          bestGain = std::max(bestGain,newBestGain);
+        }
```

Comment:
It looks like the strategy you're using here is to check if the entropy of the child is different than the entropy of the child set.  That might happen if the child node makes any additional splits.  But when we call `child->Train()`, it will return either exactly the value of `prevChildGain`, if no further splits were done (i.e. if that child was a leaf), or something better, if further splits were done.  So there should be no need to compute or store `prevChildGain`.

Also, it looks like your strategy for computing `bestGain` is to subtract the contribution of the child points from the original `bestGain` (computed at the beginning of the method), and then add back in the contribution of the newly-built child node.  I think we can ignore the original `bestGain` computation though and essentially do something like this:

```
// Outside of the loop, initialize bestGain.
bestGain = 0;

// This is the loop over the children.
for (size_t i = 0; i < numChildren; ++i)
{
  ...
  double childGain = child->Train(...);
  bestGain += double(childCounts[i]) / double(count) * childGain;
}
```

(The `double()` is to make sure we don't do integer division.  I think it's necessary but I'm not 100% sure.)

Let me know if I can clarify this idea further.

Reply:
Yes, I got that thanks for the suggestion